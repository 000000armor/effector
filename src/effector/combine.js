//@flow

import type {Store} from './unit.h'
import {createStore} from './createUnit'
import type {StateRef} from './index.h'
import {createStateRef} from './stateRef'
import {step} from './typedef'
import {getStoreState, getConfig, getNestedConfig, getGraph} from './getter'
import {is, isFunction} from './is'
import {unitObjectName} from './naming'
import {createLinkNode} from './forward'
import {throwError} from './throw'
import {isTemplateRegion} from './region'

export function combine(...args: any[]): Store<any> {
  if (args.length === 0) throwError('at least one argument required')
  let handler
  let stores
  let config
  if (getNestedConfig(args[0])) {
    config = getConfig(args[0])
    args = getNestedConfig(args[0])
  }
  const rawHandler = args[args.length - 1]
  if (isFunction(rawHandler)) {
    stores = args.slice(0, -1)
    handler = rawHandler
  } else {
    stores = args
  }

  let structStoreShape
  let shapeReady
  if (stores.length === 1) {
    const obj = stores[0]
    /*
      without edge case combine(Color, (Color) => '~')
      */
    if (!is.store(obj)) {
      /*
      case combine([R,G,B], ([R,G,B]) => '~')
      case combine({R,G,B}, ({R,G,B}) => '~')

      edge case combine([Color], ([Color]) => '~')
      edge case combine({Color}, ({Color}) => '~')

      edge case combine([R,G,B])
      edge case combine({R,G,B})

      edge case combine([Color])
      edge case combine({Color})
      */
      structStoreShape = obj
      shapeReady = true
    }
  }
  if (!shapeReady) {
    /*
    case combine(R,G,B, (R,G,B) => '~')
    */
    structStoreShape = stores
    /*
    without edge case combine(R,G,B)
    without edge case combine(Color)
    */
    if (handler) {
      handler = spreadArgs(handler)
    }
  }
  //$off
  const mergedStore = Array.isArray(structStoreShape)
    ? storeCombination(
      structStoreShape,
      list => list.slice(),
      [],
      config,
      handler,
    )
    : storeCombination(
      structStoreShape,
      obj => Object.assign({}, obj),
      {},
      config,
      handler,
    )
  return mergedStore
}

const spreadArgs = fn => list => fn(...list)

const storeCombination = (
  obj: any,
  clone: Function,
  defaultState: any,
  config?: string,
  fn?: Function,
) => {
  const stateNew = clone(defaultState)
  const store = createStore(stateNew, {
    name: config ? config : unitObjectName(obj),
  })
  const rawShape = createStateRef(stateNew)
  const isFresh = createStateRef(true)
  const {meta} = getGraph(store)
  meta.fn = fn
  meta.rawShape = rawShape
  meta.shape = obj
  const node = [
    step.check.defined(),
    step.mov({
      store: rawShape,
      to: 'a',
    }),
    //prettier-ignore
    step.filter({
      fn: (upd, {key}, {a}) => upd !== a[key],
    }),
    step.mov({
      store: isFresh,
      to: 'b',
    }),
    step.compute({
      fn(upd, {clone, key}, reg) {
        if (reg.b) {
          reg.a = clone(reg.a)
        }
        reg.a[key] = upd
      },
    }),
    step.mov({
      from: 'a',
      target: rawShape,
    }),
    step.mov({
      from: 'value',
      store: false,
      target: isFresh,
    }),
    step.barrier({priority: 'barrier'}),
    step.mov({
      from: 'value',
      store: true,
      target: isFresh,
    }),
    step.mov({store: rawShape}),
    fn && step.compute({fn}),
    step.check.changed({
      store: getStoreState(store),
    }),
  ]

  for (const key in obj) {
    const child = obj[key]
    if (!is.store(child)) {
      stateNew[key] = defaultState[key] = child
      continue
    }
    defaultState[key] = child.defaultState
    stateNew[key] = child.getState()
    createLinkNode(child, store, {
      scope: {key, clone},
      node,
      meta: {op: 'combine'},
    })
  }

  store.defaultShape = obj
  if (!isTemplateRegion()) {
    store.defaultState = fn
      ? (getStoreState(store).current = fn(stateNew))
      : defaultState
  }
  return store
}
