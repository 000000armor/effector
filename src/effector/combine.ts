import type {Store} from './unit.h'
import {createStore} from './createUnit'
import {createStateRef, addRefOp} from './stateRef'
import {step} from './typedef'
import {onConfigNesting} from './config'
import {getStoreState, setMeta} from './getter'
import {is, isFunction, isObject} from './is'
import {unitObjectName} from './naming'
import {createLinkNode} from './forward'
import {throwError} from './throw'
import {readTemplate} from './region'
import {forIn} from './collection'
import {BARRIER, MAP, REG_A, VALUE} from './tag'
import {applyTemplate} from './template'
import {callStack} from './caller'

export function combine(...args: any[]): Store<any> {
  let handler
  let stores
  let config
  onConfigNesting(args[0], (injectedData, userConfig) => {
    config = injectedData
    args = userConfig
  })
  const rawHandler = args[args.length - 1]
  if (isFunction(rawHandler)) {
    stores = args.slice(0, -1)
    handler = rawHandler
  } else {
    stores = args
  }

  let structStoreShape
  let shapeReady
  if (stores.length === 1) {
    const obj = stores[0]
    /*
      without edge case combine(Color, (Color) => '~')
      */
    if (!is.store(obj)) {
      /*
      case combine([R,G,B], ([R,G,B]) => '~')
      case combine({R,G,B}, ({R,G,B}) => '~')

      edge case combine([Color], ([Color]) => '~')
      edge case combine({Color}, ({Color}) => '~')

      edge case combine([R,G,B])
      edge case combine({R,G,B})

      edge case combine([Color])
      edge case combine({Color})
      */
      structStoreShape = obj
      shapeReady = true
    }
  }
  let noArraySpread: boolean | void
  if (!shapeReady) {
    /*
    case combine(R,G,B, (R,G,B) => '~')
    */
    structStoreShape = stores
    /*
    without edge case combine(R,G,B)
    without edge case combine(Color)
    */
    if (handler) {
      noArraySpread = true
      const fn = handler
      handler = (list: any[]) => fn(...list)
    }
  }
  if (!isObject(structStoreShape)) throwError('shape should be an object')
  return storeCombination(
    Array.isArray(structStoreShape),
    !noArraySpread,
    structStoreShape,
    config,
    handler,
  )
}

const storeCombination = (
  isArray: boolean,
  needSpread: boolean,
  obj: any,
  config?: string,
  fn?: (upd: any) => any,
) => {
  const clone = isArray ? (list: any) => list.slice() : (obj: any) => ({...obj})
  const defaultState: any = isArray ? [] : {}

  const stateNew = clone(defaultState)
  const rawShape = createStateRef(stateNew)
  const isFresh = createStateRef(true)
  rawShape.type = isArray ? 'list' : 'shape'
  rawShape.noInit = true
  applyTemplate('combineBase', rawShape, isFresh)
  const store = createStore(stateNew, {
    name: config ? config : unitObjectName(obj),
  })
  const storeStateRef = getStoreState(store)
  storeStateRef.noInit = true
  setMeta(store, 'isCombine', true)
  const node = [
    step.mov({store: rawShape, to: REG_A}),
    step.mov({store: isFresh, to: 'b'}),
    step.compute({
      safe: true,
      filter: true,
      fn(upd, {key}, reg) {
        if (upd !== reg.a[key]) {
          if (needSpread && reg.b) {
            reg.a = clone(reg.a)
          }
          reg.a[key] = upd
          return true
        }
      },
    }),
    step.mov({from: REG_A, target: rawShape}),
    step.mov({from: VALUE, store: false, target: isFresh}),
    step.mov({
      from: VALUE,
      store: true,
      target: isFresh,
      priority: BARRIER,
      batch: true,
    }),
    step.mov({store: rawShape}),
    fn && step.compute({fn: callStack}),
    step.changed({store: storeStateRef}),
  ]
  forIn(obj, (child: Store<any> | any, key) => {
    if (!is.store(child)) {
      stateNew[key] = defaultState[key] = child
      return
    }
    defaultState[key] = child.defaultState
    stateNew[key] = child.getState()
    const linkNode = createLinkNode(child, store, {
      scope: {key, fn},
      node,
      meta: {op: 'combine'},
    })
    const childRef = getStoreState(child)
    addRefOp(rawShape, {type: 'field', field: key, from: childRef})
    applyTemplate('combineField', childRef, linkNode)
  })

  store.defaultShape = obj
  addRefOp(storeStateRef, {
    type: MAP,
    from: rawShape,
    fn,
  })
  if (!readTemplate()) {
    store.defaultState = fn
      ? (storeStateRef.current = fn(stateNew))
      : defaultState
  }
  return store
}

export function createStoreObject(...args: any[]) {
  console.error('createStoreObject is deprecated, use combine instead')
  return combine(...args)
}
