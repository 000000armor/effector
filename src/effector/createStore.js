//@flow

import $$observable from 'symbol-observable'

import type {Store, Event} from './unit.h'
import {upsertLaunch} from './kernel'
import {
  own,
  step,
  createNode,
  createStateRef,
  readRef,
  bind,
  writeRef,
  is,
  getGraph,
  nextUnitID,
} from './stdlib'
import {createEvent} from './createEvent'
import {createLinkNode} from './forward'
import {createName, mapName, type CompositeName} from './naming'
import {thru} from './thru'
import type {Subscriber} from './index.h'
import {watchUnit} from './watch'
import {
  type Config,
  type StoreConfigPart as ConfigPart,
  normalizeConfig,
} from './config'
import {createSubscription} from './subscription'

export function createStore<State>(
  currentState: State,
  props: {
    +config: ConfigPart,
    +parent?: CompositeName,
    ...
  },
): Store<State> {
  const config = normalizeConfig(props)
  const id = nextUnitID()
  const {parent, sid = null, strict = true, named = null} = config
  if (strict && currentState === undefined)
    throw Error("current state can't be undefined, use null instead")
  const name = named ? named : config.name || id
  const plainState = createStateRef(currentState)
  const oldState = createStateRef(currentState)
  const compositeName = createName(name, parent)

  const updates = createEvent({named: 'updates'})

  const store: $Shape<Store<State>> = ({
    subscribers: new Map(),
    compositeName,
    graphite: createNode({
      scope: {state: plainState},
      node: [
        step.check.defined(),
        step.update({
          store: plainState,
        }),
        step.check.changed({
          store: oldState,
        }),
        step.update({
          store: oldState,
        }),
      ],
      child: [updates],
      meta: {unit: 'store', name: compositeName.shortName, sid, named},
    }),
    kind: 'store',
    id,
    shortName: compositeName.shortName,
    domainName: parent,
    updates,
    defaultConfig: config,
    defaultState: currentState,
    getState: bind(readRef, plainState),
    stateRef: plainState,
    sid,
  }: any)
  ;(store: any).watch = (store: any).subscribe = bind(watch, store)
  ;(store: any).reset = bind(reset, store)
  ;(store: any).on = bind(on, store)
  ;(store: any).off = bind(off, store)
  ;(store: any).map = bind(mapStore, store)
  ;(store: any).thru = bind(thru, store)
  ;(store: any).setState = state => {
    if (readRef(plainState) !== state) upsertLaunch([store], [state])
  }
  //$off
  store[$$observable] = bind(observable, store)
  own(store, [updates])
  return store
}

function reset(storeInstance: Store<any>, ...events: Array<Event<any>>) {
  for (const event of events)
    storeInstance.on(event, () => storeInstance.defaultState)
  return storeInstance
}

function off(storeInstance: Store<any>, event: Event<any>) {
  const currentSubscription = storeInstance.subscribers.get(event)
  if (currentSubscription !== undefined) {
    currentSubscription()
    storeInstance.subscribers.delete(event)
  }
  return storeInstance
}

function on(storeInstance: Store<any>, event: any, handler: Function) {
  storeInstance.off(event)
  storeInstance.subscribers.set(
    event,
    createSubscription(
      createLinkNode(event, storeInstance, {
        scope: {
          handler,
          state: storeInstance.stateRef,
        },
        node: [
          step.compute({
            fn(newValue, {handler, state}) {
              const result = handler(readRef(state), newValue)
              if (result === undefined) return
              return writeRef(state, result)
            },
          }),
        ],
        meta: {op: 'on'},
      }),
    ),
  )
  return storeInstance
}
const observable = (storeInstance: Store<any>) => ({
  subscribe(observer: Subscriber<any>) {
    if (observer !== Object(observer))
      throw Error('expect observer to be an object') // or function
    return watch(storeInstance, state => {
      if (observer.next) {
        observer.next(state)
      }
    })
  },
  //$off
  [$$observable]() {
    return this
  },
})
function watch(
  storeInstance: Store<any>,
  eventOrFn: Event<*> | Function,
  fn?: Function,
) {
  if (!fn || !is.unit(eventOrFn)) {
    if (typeof eventOrFn !== 'function')
      throw Error('watch requires function handler')
    eventOrFn(storeInstance.getState())
    return watchUnit(storeInstance, eventOrFn)
  }
  if (typeof fn !== 'function')
    throw Error('second argument should be a function')
  return eventOrFn.watch(payload => fn(storeInstance.getState(), payload))
}

function mapStore<A, B>(
  store: Store<A>,
  fn: (state: A, lastState?: B) => B,
  firstState?: B,
): Store<B> {
  let config
  let name
  if (typeof fn === 'object') {
    config = fn
    name = fn.name
    firstState = fn.firstState
    fn = fn.fn
  }
  let lastResult
  const storeState = store.getState()
  if (storeState !== undefined) {
    lastResult = fn(storeState, firstState)
  }

  const innerStore: Store<any> = createStore(lastResult, {
    name: mapName(store, name),
    parent: store.domainName,
    config,
    strict: false,
  })
  createLinkNode(store, innerStore, {
    scope: {
      handler: fn,
      state: innerStore.stateRef,
    },
    node: [
      step.compute({
        fn: (upd, {state, handler}) => handler(upd, readRef(state)),
      }),
      step.check.defined(),
      step.check.changed({
        store: innerStore.stateRef,
      }),
    ],
    meta: {op: 'map'},
  })
  return innerStore
}
