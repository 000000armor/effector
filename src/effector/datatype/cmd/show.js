//@flow

import * as Name from './type'
import * as Type from './index.h'

export function show(value: Type.Cmd): string {
 switch (value.type) {
  case Name.EMIT:
   return `Emit(${value.data.fullName})`
  case Name.RUN:
   return 'Run'
  case Name.COMPUTE:
   return 'Compute'
  case Name.FILTER:
   return 'Filter'
  default:
   /*::(value.type: empty)*/
   throw new Error('impossible type')
 }
}
