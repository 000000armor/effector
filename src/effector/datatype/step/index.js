//@flow

export * from './type.h'
export * from './type'
export * from './index.h'
export * from './impl'
