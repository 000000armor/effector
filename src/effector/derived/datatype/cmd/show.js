//@flow

import * as Name from './type.h'
import * as Type from './index.h'

export function show(value: Type.Cmd): string {
 switch (value.type) {
  case ('emit': Name.EmitType):
   return `Emit(${value.data.fullName})`
  case ('run': Name.RunType):
   return 'Run'
  case ('compute': Name.ComputeType):
   return 'Compute'
  default:
   /*::(value.type: empty)*/
   throw new Error('impossible type')
 }
}
