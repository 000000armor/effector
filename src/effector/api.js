//@flow

// import invariant from 'invariant'
import $$observable from 'symbol-observable'
import warning from 'warning'
import {from, type Stream} from 'most'
import type {Event, Effect, GraphiteMeta} from './index.h'
import * as Kind from '../kind'
import {setProperty} from '../setProperty'
// import {runSyncGraph} from '../../graph'
import * as Cmd from './datatype/cmd'
import * as Step from './datatype/step'
import {walkEvent} from './walk'

export function createEvent<Payload>(name: string): Event<Payload> {
 return eventConstructor({name, domainName: ''})
}

export function createEffect<Payload, Done>(
 name: string,
): Effect<Payload, Done, *> {
 return effectConstructor({name, domainName: ''})
}

function eventConstructor<Payload>({
 name,
 domainName,
}: {
 name: string,
 domainName: string,
}): Event<Payload> {
 const fullName = makeName(name, domainName)

 const cmd: Cmd.Emit = Cmd.emit({
  subtype: 'event',
  fullName,
  runner: createGraphite,
 })
 const step: Step.Single = Step.single(cmd)
 const nextSteps: Step.Multi = Step.multi()
 const stepFull: Step.Seq = Step.seq([step, nextSteps])
 const graphite: GraphiteMeta = {
  cmd,
  step,
  next: nextSteps,
  seq: stepFull,
 }
 const instance = (payload: Payload): Payload =>
  instanceAsEvent.create(payload, fullName)
 function createGraphite(payload: Payload): Payload {
  return instanceAsEvent.create(payload, fullName)
 }
 const instanceAsEvent: Event<Payload> = (instance: any)
 instanceAsEvent.graphite = graphite

 setProperty('create', create, instance)

 setProperty('toString', getType, instance)
 setProperty('getType', getType, instance)
 setProperty('kind', Kind.EVENT, instance)
 setProperty($$observable, () => instance, instance)
 instance.watch = watch
 instance.map = map
 instance.prepend = prepend
 instance.subscribe = subscribe
 instance.to = to
 instance.epic = epic
 // instance.link = (b, ab, ba) => link(instanceAsEvent, b, ab, ba)
 function epic<T>(fn: (Stream<Payload>) => Stream<T>): Event<T> {
  const instance$ = from(instanceAsEvent).multicast()
  const epic$ = fn(instance$).multicast()
  const mapped = eventConstructor({name: `${name}$ ~> *`, domainName})
  epic$.observe(e => {
   mapped.create(e, fullName)
  })
  return mapped
 }
 function to(target, handler?: Function) {
  switch (Kind.readKind(target)) {
   case Kind.STORE:
    return watch(payload => target.setState(payload, handler))
   case Kind.EVENT:
   case Kind.EFFECT:
    return watch(target.create)
   default: {
    throw new TypeError(`Unsupported kind`)
   }
  }
 }
 const seq = 0

 function create(payload, fullName) {
  walkEvent(payload, instanceAsEvent)
  return payload
 }
 function watch(watcher: (payload: Payload, type: string) => any) {
  const runCmd = Step.single(
   Cmd.run({
    runner(newValue: Payload) {
     return watcher(newValue, fullName)
    },
   }),
  )
  instanceAsEvent.graphite.next.data.add(runCmd)
  return () => {
   instanceAsEvent.graphite.next.data.delete(runCmd)
  }
 }

 function subscribe(observer) {
  const unsub = watch(payload => observer.next(payload))
  function unsubscribe() {
   unsub()
  }

  unsubscribe.unsubscribe = unsubscribe

  return unsubscribe
 }
 function map<Next>(fn: Payload => Next) {
  const mapped = eventConstructor({name: `${name} → *`, domainName})
  const computeCmd = Step.single(
   Cmd.compute({
    reduce(_, newValue: Payload, ctx) {
     return fn(newValue)
    },
   }),
  )
  const nextSeq = Step.seq([computeCmd, ...mapped.graphite.seq.data])
  instanceAsEvent.graphite.next.data.add(nextSeq)
  return mapped
 }
 function prepend<Before>(fn: Before => Payload) {
  const contramapped: Event<Before> = eventConstructor({
   name: `* → ${name}`,
   domainName,
  })

  const computeCmd = Step.single(
   Cmd.compute({
    reduce(_, newValue: Before, ctx) {
     return fn(newValue)
    },
   }),
  )
  const nextSeq = Step.seq([computeCmd, ...instanceAsEvent.graphite.seq.data])
  contramapped.graphite.next.data.add(nextSeq)
  return contramapped
 }
 function getType() {
  return fullName
 }
 return (instance: $todo)
}

function effectConstructor<Payload, Done>({
 name,
 domainName,
}: {
 name: string,
 domainName: string,
}): Effect<Payload, Done, *> {
 const instanceAsEvent: Event<Payload> = eventConstructor({
  name,
  domainName,
 })
 const instance: Effect<Payload, Done, any> = (instanceAsEvent: any)
 const eventCreate = instanceAsEvent.create
 const done: Event<{params: Payload, result: Done}> = eventConstructor({
  name: `${name} done`,
  domainName,
 })
 const fail: Event<{params: Payload, error: *}> = eventConstructor({
  name: `${name} fail`,
  domainName,
 })

 // instanceAsEvent.step.data.delete(instanceAsEvent.cmd)
 instance.done = done
 instance.fail = fail
 instance.use = use
 setProperty('create', create, instance)
 setProperty('kind', Kind.EFFECT, instance)
 let thunk = async(value: Payload): Promise<Done> => {
  warning(false, 'no thunk used')
  declare var result: Done
  /*::return result*/
 }
 function use(fn) {
  thunk = fn
 }
 function create(payload: Payload) {
  const throwSymbol = {}
  let syncError = throwSymbol
  let req
  try {
   req = thunk(payload)
  } catch (err) {
   syncError = err
  }
  const isSyncError = syncError !== throwSymbol
  const isPromise = !isSyncError && hasPromise(req)
  //$todo
  const onDone = result => done({params: payload, result})
  const onFail = error => fail({params: payload, error})
  //$todo
  if (isPromise) req.then(onDone, onFail)
  else if (isSyncError) onFail(syncError)
  else onDone(req)
  eventCreate(payload, instanceAsEvent.getType())
  return {
   done() {
    return new Promise(rs => {
     if (isPromise) req.then(result => rs({params: payload, result}))
     else if (!isSyncError) rs({params: payload, result: req})
    })
   },
   fail() {
    return new Promise(rs => {
     if (isPromise) req.catch(error => rs({params: payload, error}))
     else if (isSyncError) rs({params: payload, error: syncError})
    })
   },
   promise() {
    return new Promise((rs, rj) => {
     if (isPromise) req.then(rs, rj)
     else if (isSyncError) rj(syncError)
     else rs(req)
    })
   },
  }
 }

 return instance
}

function hasPromise(req: mixed): boolean %checks {
 return (
  typeof req === 'object' && req !== null && typeof req.then === 'function'
 )
}

function makeName(name, domainName) {
 return [name, domainName].filter(str => str.length > 0).join('/')
}
