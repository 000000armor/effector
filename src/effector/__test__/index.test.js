//@flow

import * as Api from '../api'
import {createEvent} from '../api'
import {combine} from '../combine'
import * as Type from '../index.h'
import {createStore} from '../../store/createStore'
import {createStoreObject} from '../../store/createStoreObject'
import {show} from '../datatype/step/show'

test('graphite', () => {
 const fn = jest.fn()
 const fn1 = jest.fn()
 const foo: Type.Event<number> = Api.createEvent('foo')
 const bar = foo.map(x => (fn(x), x + 1))
 const store1: Type.Store<string> = createStore('foo').on(bar, (state, bar) =>
  [state, bar].join(' | '),
 )
 const store2 = store1.map(e => e.length)
 const unsub1 = bar.watch(fn1)
 foo(0)
 foo(10)
 expect(fn).toHaveBeenCalledTimes(2)
 expect(store1.getState()).toBe('foo | 1 | 11')
 expect(foo.graphite.seq).toMatchSnapshot('foo seq')
 expect(bar.graphite.seq).toMatchSnapshot('bar seq')
 const showBar = show(bar.graphite.seq)
 console.log(showBar)

 expect(showBar).toMatchSnapshot('show bar')
 unsub1()
 const showBar2 = show(bar.graphite.seq)
 console.log(showBar2)
 expect(showBar2).toMatchSnapshot('show bar')
 expect(bar.graphite.seq).toMatchSnapshot('bar seq')
 foo(100)
 expect(fn1).toHaveBeenCalledTimes(2)
 expect(store1.graphite.seq).toMatchSnapshot('store1 seq')
 expect(store2.graphite.seq).toMatchSnapshot('store2 seq')
 const showFoo = show(foo.graphite.seq)
 const showStore = show(store1.graphite.seq)
 const showStore2 = show(store2.graphite.seq)
 console.log(showFoo)
 console.log(showStore)
 console.log(showStore2)
 expect(showFoo).toMatchSnapshot('show foo')
 expect(showStore).toMatchSnapshot('show store1')
 expect(showStore2).toMatchSnapshot('show store2')
})

test('showcase', () => {
 const fn = jest.fn()
 const foo = createEvent('foo')
 const bar = createEvent('bar')

 const a = createStore(1)
 const b = createStore(2)
 const bigStore = createStoreObject({a, b})
 const mapped = bigStore.map(s => s.a)

 a.on(foo, n => n + 1)
 b.on(bar, n => n + 1)

 mapped.watch(fn)

 foo()
 foo()

 bar()
 bar()
 bar()
 bar()

 expect(show(a.graphite.seq)).toMatchSnapshot('store a')
 expect(show(foo.graphite.seq)).toMatchSnapshot('event foo')
 expect(show(mapped.graphite.seq)).toMatchSnapshot('mapped')
 expect(fn).toHaveBeenCalledTimes(3)
 const first = createStore('s')
 const second = createStore('h')
 const third = createStore('i')
 const status = combine(first, second, third, (a, b, c) => [a, b, c].join(''))
 expect(status.getState()).toBe('shi')
})
