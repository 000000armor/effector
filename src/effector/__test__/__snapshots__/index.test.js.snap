// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphite: bar seq 1`] = `
Seq {
  "data": Array [
    Single {
      "data": Emit {
        "data": Object {
          "fullName": "foo → *",
          "runner": [Function],
          "subtype": "event",
        },
      },
    },
    Multi {
      "data": Set {
        Seq {
          "data": Array [
            Single {
              "data": Compute {
                "data": Object {
                  "reduce": [Function],
                },
              },
            },
            Single {
              "data": Filter {
                "data": Object {
                  "filter": [Function],
                },
              },
            },
            Single {
              "data": Filter {
                "data": Object {
                  "filter": [Function],
                },
              },
            },
            Single {
              "data": Compute {
                "data": Object {
                  "reduce": [Function],
                },
              },
            },
            Multi {
              "data": Set {
                Seq {
                  "data": Array [
                    Single {
                      "data": Filter {
                        "data": Object {
                          "filter": [Function],
                        },
                      },
                    },
                    Single {
                      "data": Compute {
                        "data": Object {
                          "reduce": [Function],
                        },
                      },
                    },
                    Single {
                      "data": Filter {
                        "data": Object {
                          "filter": [Function],
                        },
                      },
                    },
                    Single {
                      "data": Filter {
                        "data": Object {
                          "filter": [Function],
                        },
                      },
                    },
                    Single {
                      "data": Compute {
                        "data": Object {
                          "reduce": [Function],
                        },
                      },
                    },
                    Multi {
                      "data": Set {},
                    },
                  ],
                },
              },
            },
          ],
        },
        Single {
          "data": Run {
            "data": Object {
              "runner": [Function],
            },
          },
        },
      },
    },
  ],
}
`;

exports[`graphite: bar seq 2`] = `
Seq {
  "data": Array [
    Single {
      "data": Emit {
        "data": Object {
          "fullName": "foo → *",
          "runner": [Function],
          "subtype": "event",
        },
      },
    },
    Multi {
      "data": Set {
        Seq {
          "data": Array [
            Single {
              "data": Compute {
                "data": Object {
                  "reduce": [Function],
                },
              },
            },
            Single {
              "data": Filter {
                "data": Object {
                  "filter": [Function],
                },
              },
            },
            Single {
              "data": Filter {
                "data": Object {
                  "filter": [Function],
                },
              },
            },
            Single {
              "data": Compute {
                "data": Object {
                  "reduce": [Function],
                },
              },
            },
            Multi {
              "data": Set {
                Seq {
                  "data": Array [
                    Single {
                      "data": Filter {
                        "data": Object {
                          "filter": [Function],
                        },
                      },
                    },
                    Single {
                      "data": Compute {
                        "data": Object {
                          "reduce": [Function],
                        },
                      },
                    },
                    Single {
                      "data": Filter {
                        "data": Object {
                          "filter": [Function],
                        },
                      },
                    },
                    Single {
                      "data": Filter {
                        "data": Object {
                          "filter": [Function],
                        },
                      },
                    },
                    Single {
                      "data": Compute {
                        "data": Object {
                          "reduce": [Function],
                        },
                      },
                    },
                    Multi {
                      "data": Set {},
                    },
                  ],
                },
              },
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`graphite: foo seq 1`] = `
Seq {
  "data": Array [
    Single {
      "data": Emit {
        "data": Object {
          "fullName": "foo",
          "runner": [Function],
          "subtype": "event",
        },
      },
    },
    Multi {
      "data": Set {
        Seq {
          "data": Array [
            Single {
              "data": Compute {
                "data": Object {
                  "reduce": [Function],
                },
              },
            },
            Single {
              "data": Emit {
                "data": Object {
                  "fullName": "foo → *",
                  "runner": [Function],
                  "subtype": "event",
                },
              },
            },
            Multi {
              "data": Set {
                Seq {
                  "data": Array [
                    Single {
                      "data": Compute {
                        "data": Object {
                          "reduce": [Function],
                        },
                      },
                    },
                    Single {
                      "data": Filter {
                        "data": Object {
                          "filter": [Function],
                        },
                      },
                    },
                    Single {
                      "data": Filter {
                        "data": Object {
                          "filter": [Function],
                        },
                      },
                    },
                    Single {
                      "data": Compute {
                        "data": Object {
                          "reduce": [Function],
                        },
                      },
                    },
                    Multi {
                      "data": Set {
                        Seq {
                          "data": Array [
                            Single {
                              "data": Filter {
                                "data": Object {
                                  "filter": [Function],
                                },
                              },
                            },
                            Single {
                              "data": Compute {
                                "data": Object {
                                  "reduce": [Function],
                                },
                              },
                            },
                            Single {
                              "data": Filter {
                                "data": Object {
                                  "filter": [Function],
                                },
                              },
                            },
                            Single {
                              "data": Filter {
                                "data": Object {
                                  "filter": [Function],
                                },
                              },
                            },
                            Single {
                              "data": Compute {
                                "data": Object {
                                  "reduce": [Function],
                                },
                              },
                            },
                            Multi {
                              "data": Set {},
                            },
                          ],
                        },
                      },
                    },
                  ],
                },
                Single {
                  "data": Run {
                    "data": Object {
                      "runner": [Function],
                    },
                  },
                },
              },
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`graphite: show bar 1`] = `
"Seq [
  1. Single Emit(foo → *),
  2. Multi [
     *  Seq [
        1. Single Compute,
        2. Single Filter,
        3. Single Filter,
        4. Single Compute,
        5. Multi [
           *  Seq [
              1. Single Filter,
              2. Single Compute,
              3. Single Filter,
              4. Single Filter,
              5. Single Compute,
              6. Multi []
           ]
        ]
     ],
     *  Single Run
  ]
]"
`;

exports[`graphite: show bar 2`] = `
"Seq [
  1. Single Emit(foo → *),
  2. Multi [
     *  Seq [
        1. Single Compute,
        2. Single Filter,
        3. Single Filter,
        4. Single Compute,
        5. Multi [
           *  Seq [
              1. Single Filter,
              2. Single Compute,
              3. Single Filter,
              4. Single Filter,
              5. Single Compute,
              6. Multi []
           ]
        ]
     ]
  ]
]"
`;

exports[`graphite: show foo 1`] = `
"Seq [
  1. Single Emit(foo),
  2. Multi [
     *  Seq [
        1. Single Compute,
        2. Single Emit(foo → *),
        3. Multi [
           *  Seq [
              1. Single Compute,
              2. Single Filter,
              3. Single Filter,
              4. Single Compute,
              5. Multi [
                 *  Seq [
                    1. Single Filter,
                    2. Single Compute,
                    3. Single Filter,
                    4. Single Filter,
                    5. Single Compute,
                    6. Multi []
                 ]
              ]
           ]
        ]
     ]
  ]
]"
`;

exports[`graphite: show store1 1`] = `
"Seq [
  1. Single Filter,
  2. Single Compute,
  3. Multi [
     *  Seq [
        1. Single Filter,
        2. Single Compute,
        3. Single Filter,
        4. Single Filter,
        5. Single Compute,
        6. Multi []
     ]
  ]
]"
`;

exports[`graphite: show store2 1`] = `
"Seq [
  1. Single Filter,
  2. Single Compute,
  3. Multi []
]"
`;

exports[`graphite: store1 seq 1`] = `
Seq {
  "data": Array [
    Single {
      "data": Filter {
        "data": Object {
          "filter": [Function],
        },
      },
    },
    Single {
      "data": Compute {
        "data": Object {
          "reduce": [Function],
        },
      },
    },
    Multi {
      "data": Set {
        Seq {
          "data": Array [
            Single {
              "data": Filter {
                "data": Object {
                  "filter": [Function],
                },
              },
            },
            Single {
              "data": Compute {
                "data": Object {
                  "reduce": [Function],
                },
              },
            },
            Single {
              "data": Filter {
                "data": Object {
                  "filter": [Function],
                },
              },
            },
            Single {
              "data": Filter {
                "data": Object {
                  "filter": [Function],
                },
              },
            },
            Single {
              "data": Compute {
                "data": Object {
                  "reduce": [Function],
                },
              },
            },
            Multi {
              "data": Set {},
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`graphite: store2 seq 1`] = `
Seq {
  "data": Array [
    Single {
      "data": Filter {
        "data": Object {
          "filter": [Function],
        },
      },
    },
    Single {
      "data": Compute {
        "data": Object {
          "reduce": [Function],
        },
      },
    },
    Multi {
      "data": Set {},
    },
  ],
}
`;

exports[`showcase: event foo 1`] = `
"Seq [
  1. Single Emit(foo),
  2. Multi [
     *  Seq [
        1. Single Compute,
        2. Single Filter,
        3. Single Filter,
        4. Single Compute,
        5. Multi [
           *  Single Run
        ]
     ]
  ]
]"
`;

exports[`showcase: mapped 1`] = `
"Seq [
  1. Single Filter,
  2. Single Compute,
  3. Multi [
     *  Single Run
  ]
]"
`;

exports[`showcase: store a 1`] = `
"Seq [
  1. Single Filter,
  2. Single Compute,
  3. Multi [
     *  Single Run
  ]
]"
`;
